### PR TITLE
Feature: Asymmetric Key Auth

### PR DESCRIPTION
**Pull request by dpatsora**

### CODE REVIEW

File: api/user.go [hunk 1]
```diff
@@ -33,6 +33,8 @@ func (a *API) UserGet(w http.ResponseWriter, r *http.Request) error {
 	}
 
 	aud := a.requestAud(ctx, r)
+	fmt.Println("Request aud:", aud)
+	fmt.Printf("Claims aud: %+v\n", claims)
```
Comment (dpatsora) [inline api/user.go:37]: Remove

File: migrations/20211006231400_create_keys_table.up.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	challenge_token uuid NOT NULL,
+	challenge_token_issued_at timestamptz NOT NULL,
+	challenge_token_expires_at timestamptz NOT NULL,
+	challenge_passed bool DEFAULT false NOT NULL,
+
+    created_at timestamptz NULL,
+	updated_at timestamptz NULL,
+	CONSTRAINT asymmetric_keys_pkey PRIMARY KEY (id),
+	CONSTRAINT asymmetric_keys_user_id_fkey FOREIGN KEY (user_id) REFERENCES auth.users(id) ON DELETE CASCADE
+);
```
Comment (dpatsora) [inline migrations/20211006231400_create_keys_table.up.sql:16]: new line

File: models/asymmetric_key.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/gofrs/uuid"
+	"github.com/netlify/gotrue/storage"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+const challengeExpirationDuration = 30 * time.Minute
```
Comment (dpatsora) [inline models/asymmetric_key.go:17]: challengeTokenExpirationDuration

File: models/asymmetric_key.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	return err
+}
+
+func (a *AsymmetricKey) verifyEthKeySignature(rawSignature string) error {
+	log := logrus.WithField("component", "signatureVerification")
+	log.Infof("Signature: %v", rawSignature)
+	log.Infof("Challenge token: %v", a.ChallengeToken.String())
+
+	log.Infof("Equal tokens strings: %v", a.ChallengeToken.String() == "c2918781-fa53-44b1-8a0f-7da323e6fcec")
+	log.Infof("Equal tokens bytes: %v", bytes.Equal([]byte(a.ChallengeToken.String()), []byte("c2918781-fa53-44b1-8a0f-7da323e6fcec")))
```
Comment (dpatsora) [inline models/asymmetric_key.go:124]: Remove

File: models/asymmetric_key.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	signature, err := hexutil.Decode(rawSignature)
+	if err != nil {
+		return err
+	}
+	log.Infof("Equal signatures: %v", rawSignature == "0x84be2bca9f4ab3ffa5158f0ab857987ad5e286a4b32f98c22c6d74659cff9726777ca74f7469c03b08301cbe3a1c381ece2c57249cfc6e4a0584dcff4b28d73b1c")
```
Comment (dpatsora) [inline models/asymmetric_key.go:129]: Remove

File: models/asymmetric_key.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+	// https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L442
```
Comment (dpatsora) [inline models/asymmetric_key.go:131]: Add more comments

File: models/asymmetric_key.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+	log.Infof("Address: %s", addr.String())
+	if addr.String() != a.Key {
+		return WrongPublicKeyError
+	}
+
+	return nil
+}
+
+// verifyKeyAndAlgorithm verifies public key format for specific algorithm.
+// If key satisfies conditions, nil error is returned
```
Comment (dpatsora) [inline models/asymmetric_key.go:152]: nil is returned

File: models/errors.go [hunk 8]
```diff
@@ -47,6 +49,13 @@ func (e RefreshTokenNotFoundError) Error() string {
 	return "Refresh Token not found"
 }
 
+// AsymmetricKeyNotFoundError represents when a refresh token is not found.
```
Comment (dpatsora) [inline models/errors.go:52]: Remove

**Review by mnaichuk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**